cmd/contour/serve.go

```diff
@@ -210,37 +221,46 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	// wrap eventRecorder in a converter for objects from the dynamic client.
 	dynamicHandler := &k8s.DynamicClientHandler{
 		Next:      eventRecorder,
-		Converter: k8s.NewUnstructuredConverter(),
+		Converter: converter,
 		Logger:    log.WithField("context", "dynamicHandler"),
 	}
 
 	// step 4. register our resource event handler with the k8s informers.
 	var informers []cache.SharedIndexInformer
 
-	informers = registerEventHandler(informers, dynamicInformers.ForResource(ingressroutev1.IngressRouteGVR).Informer(), dynamicHandler)
-	informers = registerEventHandler(informers, dynamicInformers.ForResource(ingressroutev1.TLSCertificateDelegationGVR).Informer(), dynamicHandler)
-	informers = registerEventHandler(informers, dynamicInformers.ForResource(projectcontour.HTTPProxyGVR).Informer(), dynamicHandler)
-	informers = registerEventHandler(informers, dynamicInformers.ForResource(projectcontour.TLSCertificateDelegationGVR).Informer(), dynamicHandler)
-	informers = registerEventHandler(informers, coreInformers.Core().V1().Services().Informer(), eventRecorder)
+	informers = registerEventHandler(informers, dynamicHandler, dynamicInformers.ForResource(ingressroutev1.IngressRouteGVR).Informer(),
+		dynamicInformers.ForResource(ingressroutev1.TLSCertificateDelegationGVR).Informer(),
+		dynamicInformers.ForResource(projectcontour.HTTPProxyGVR).Informer(),
+		dynamicInformers.ForResource(projectcontour.TLSCertificateDelegationGVR).Informer())
+
+	informers = registerEventHandler(informers, eventRecorder, coreInformers.Core().V1().Services().Informer())
+
+	if ctx.UseServiceAPITypes {
+		log.Info("Enabling Experimental Service APIs types")
+		informers = registerEventHandler(informers, dynamicHandler, dynamicInformers.ForResource(serviceapis.GroupVersion.WithResource("gatewayclasses")).Informer(),
+			dynamicInformers.ForResource(serviceapis.GroupVersion.WithResource("gateways")).Informer(),
+			dynamicInformers.ForResource(serviceapis.GroupVersion.WithResource("httproutes")).Informer(),
+			dynamicInformers.ForResource(serviceapis.GroupVersion.WithResource("tcproutes")).Informer())
+	}
 
 	// After K8s 1.13 the API server will automatically translate extensions/v1beta1.Ingress objects
 	// to networking/v1beta1.Ingress objects so we should only listen for one type or the other.
 	// The default behavior is to listen for networking/v1beta1.Ingress objects and let the API server
 	// transparently upgrade the extensions version for us.
 	if ctx.UseExtensionsV1beta1Ingress {
-		informers = registerEventHandler(informers, coreInformers.Extensions().V1beta1().Ingresses().Informer(), eventRecorder)
+		informers = registerEventHandler(informers, eventRecorder, coreInformers.Extensions().V1beta1().Ingresses().Informer())
```

Comment:
why are the last two arguments reversed? This smells like the api of registerEventHandler is hard to use

Reply:
This was a request from @jpeach earlier.

Reply:
oh, my mistake. Is it possible for you to pull this change into another PR so we're not arguing about the color of the drapes.

Reply:
I think I'm going to back this change out from here, and log an issue about it, I think the problem is deeper than the readability having a lot of `informers =` lines.

Reply:
agreed, the solution is probably to break this out into a function of its own

Reply:
Yep, PR incoming for that.